File: examples/integrators/pattern_adaptive_step.py
#!/usr/bin/env python3
"""
Simple adaptive-step integration pattern example.

Demonstrates the basic pattern for using an adaptive-step integrator
with exponential decay dynamics.
"""

import brahe as bh
import numpy as np


def dynamics(t, state):
    """Exponential decay dynamics: dx/dt = -k*x"""
    k = 0.1
    return np.array([-k * state[0]])


# Create adaptive integrator
config = bh.IntegratorConfig.adaptive(abs_tol=1e-10, rel_tol=1e-9)
integrator = bh.DP54Integrator(1, dynamics, config=config)

# Integrate with automatic step control
t = 0.0
initial_state = np.array([1.0])
dt = 60.0  # Initial guess

result = integrator.step(t, initial_state, dt)

print(f"Initial state: {initial_state[0]:.6f}")
print(f"State after step: {result.state[0]:.6f}")
print(f"Step used: {result.dt_used:.2f}s")
print(f"Recommended next step: {result.dt_next:.2f}s")
print(f"Error estimate: {result.error_estimate:.2e}")

File: examples/integrators/pattern_adaptive_step.rs
//! ```cargo
//! [dependencies]
//! brahe = { path = "../.." }
//! nalgebra = "0.34"
//! ```
//!
//! Simple adaptive-step integration pattern example.
//!
//! Demonstrates the basic pattern for using an adaptive-step integrator
//! with exponential decay dynamics.

use brahe::integrators::*;
use nalgebra::DVector;

fn main() {
    // Dynamics function: Exponential decay dx/dt = -k*x
    let k = 0.1;
    let dynamics = move |_t: f64, state: DVector<f64>, _params: Option<&DVector<f64>>| -> DVector<f64> {
        DVector::from_vec(vec![-k * state[0]])
    };

    // Create adaptive integrator
    let config = IntegratorConfig::adaptive(1e-10, 1e-9);
    let integrator = DormandPrince54DIntegrator::with_config(
        1,
        Box::new(dynamics),
        None,
        None,
        None,
        config
    );

    // Integrate with automatic step control
    let t = 0.0;
    let initial_state = DVector::from_vec(vec![1.0]);
    let dt = 60.0; // Initial guess

    let result = integrator.step(t, initial_state.clone(), Some(dt));

    println!("Initial state: {:.6}", initial_state[0]);
    println!("State after step: {:.6}", result.state[0]);
    println!("Step used: {:.2}s", result.dt_used);
    println!("Recommended next step: {:.2}s", result.dt_next);
    println!("Error estimate: {:.2e}", result.error_estimate.unwrap());
}
